source/agora/consensus/protocol/Nominator.d

```diff
@@ -750,21 +756,28 @@ extern(D):
         const Scalar block_challenge = block_hash;
         // Fetch the R from enrollment commitment for signing validator
         const CR = this.enroll_man.getCommitmentNonce(block_sig.public_key, block_sig.height);
+        // Get the preimage at ledger height
+        const preimage = this.enroll_man.getValidatorPreimageAt(block_sig.public_key, Height(block_sig.height - 1));
+        if (preimage == Hash.init)
+        {
+            log.warn("No preimage for {} to sign block at height {}", block_sig.public_key, block_sig.height);
+            return false;
+        }
         // Determine the R of signature (R, s)
-        const Point R = CR + block_challenge.toPoint();
-        log.trace("collectBlockSignature: [{}] Enrollment commitment (CR): {}, signing commitment (R): {}",
-                  block_sig.public_key, CR, R);
+        Point R = CR + Scalar(preimage).toPoint();
         // Compose the signature (R, s)
         const sig = Signature(R, block_sig.signature.asScalar());
         // Check this signature is valid for this block and signing validator
         if (!verify(sig, block_challenge, K))
         {
-            log.warn("collectBlockSignature: INVALID Block signature received for slot {} from node {}",
-                block_sig.height, block_sig.public_key);
+            log.warn("collectBlockSignature height {} node {}: INVALID Block signature " ~
+                "Enrollment commitment (CR): {}, signing commitment (R): {}, preimage: {}",
+                    block_sig.height, block_sig.public_key, CR, R, preimage);
             return false;
         }
-        log.trace("collectBlockSignature: VALID block signature at height {} for node {}",
-            block_sig.height, block_sig.public_key);
+        log.trace("collectBlockSignature height {} node {}: VALID block signature " ~
+            "- Enrollment commitment (CR): {}, signing commitment (R): {}, preimage: {}",
+                block_sig.height, block_sig.public_key, CR, R, preimage);
```

Comment:
Can we remove this ?

Reply:
Or shorten it greatly, to the tune of "Accepted signature" + height + validator key

Reply:
Removed